File: Tools/event_scoring.py
#!/usr/bin/env python3
"""
Regra de classificação de uma prova (referência; Kotlin e Swift portam isto).

É a única parte do produto que não vem da física da medição, mas ainda assim precisa ser idêntica nas
duas plataformas — daí morar aqui e ser conferida por um vetor compartilhado (`event_ranking.json`).

Regra:
  tempo final = tempo REFINADO da passada + 5 s por tambor derrubado;
  "sem tempo" (SAT) fica sempre por último, sem colocação;
  empate resolve pelo tempo bruto e, persistindo, pela ordem de largada.

O tempo refinado é sempre o usado (mesmo em qualidade 0, onde ele é o centro do intervalo físico do
gatilho): é o mesmo número que o app mostra na tela, e classificar por outro seria mentir para o
competidor.
"""
from dataclasses import dataclass, field
from typing import List, Optional

PENALTY_PER_BARREL_NS = 5_000_000_000


@dataclass
class Run:
    """O mínimo de uma passada para classificar (espelha RunRecord nas duas plataformas)."""
    entry_order: int                 # número de largada
    elapsed_refined_ns: int
    elapsed_raw_ns: int
    barrels_knocked: int = 0
    no_time: bool = False
    category: str = ""
    rider: str = ""


@dataclass
class Placing:
    entry_order: int
    place: Optional[int]             # None para SAT
    final_ns: int
    penalty_ns: int


def penalty_ns(run: Run) -> int:
    return run.barrels_knocked * PENALTY_PER_BARREL_NS


def final_ns(run: Run) -> int:
    """Tempo que vale para a classificação (refinado + penalidade). SAT não tem tempo válido."""
    return run.elapsed_refined_ns + penalty_ns(run)


def rank(runs: List[Run]) -> List[Placing]:
    """
    Classifica uma lista de passadas (já filtrada por categoria, se for o caso). Ordem estável e
    determinística nas três linguagens: SAT por último; depois tempo final, tempo bruto, ordem de
    largada.
    """
    def key(r: Run):
        return (1 if r.no_time else 0, final_ns(r), r.elapsed_raw_ns, r.entry_order)

    out: List[Placing] = []
    place = 0
    for r in sorted(runs, key=key):
        if r.no_time:
            out.append(Placing(r.entry_order, None, 0, penalty_ns(r)))
        else:
            place += 1
            out.append(Placing(r.entry_order, place, final_ns(r), penalty_ns(r)))
    return out


def rank_by_category(runs: List[Run]) -> List[Placing]:
    """Classifica dentro de cada categoria; a saída sai agrupada por categoria (ordem alfabética)."""
    out: List[Placing] = []
    for cat in sorted({r.category for r in runs}):
        out.extend(rank([r for r in runs if r.category == cat]))
    return out

File: web/src/core/eventScoring.ts
import type { Nanos } from "./nanos.ts";

/**
 * Regra de classificação de uma prova (porte de `Tools/event_scoring.py`, conferida pelo vetor
 * compartilhado `event_ranking.json` — a mesma dos núcleos Kotlin e Swift).
 *
 * Tempo final = tempo REFINADO da passada + 5 s por tambor derrubado; "sem tempo" (SAT) fica sempre
 * por último, sem colocação; empate resolve pelo tempo bruto e, persistindo, pela ordem de largada.
 */
export const PENALTY_PER_BARREL_NS = 5_000_000_000;

/** O mínimo de uma passada para classificar. */
export interface ScoringRun {
  entryOrder: number;
  elapsedRefinedNs: Nanos;
  elapsedRawNs: Nanos;
  barrelsKnocked: number;
  noTime: boolean;
  category: string;
}

export interface Placing {
  entryOrder: number;
  /** null para SAT (sem colocação). */
  place: number | null;
  finalNs: Nanos;
  penaltyNs: Nanos;
}

export const penaltyNs = (r: ScoringRun): Nanos => r.barrelsKnocked * PENALTY_PER_BARREL_NS;

export const finalNs = (r: ScoringRun): Nanos => r.elapsedRefinedNs + penaltyNs(r);

/** Classifica uma lista já filtrada por categoria. Ordem determinística nas quatro linguagens. */
export function rank(runs: ScoringRun[]): Placing[] {
  // o índice de entrada entra como último desempate: `Array.sort` do JS é estável, mas explicitar
  // mantém a regra idêntica à do Swift (cuja ordenação não é) e à do Kotlin
  const sorted = runs
    .map((r, i) => ({ r, i }))
    .sort((a, b) => {
      const na = a.r.noTime ? 1 : 0;
      const nb = b.r.noTime ? 1 : 0;
      if (na !== nb) return na - nb;
      const fa = finalNs(a.r);
      const fb = finalNs(b.r);
      if (fa !== fb) return fa - fb;
      if (a.r.elapsedRawNs !== b.r.elapsedRawNs) return a.r.elapsedRawNs - b.r.elapsedRawNs;
      if (a.r.entryOrder !== b.r.entryOrder) return a.r.entryOrder - b.r.entryOrder;
      return a.i - b.i;
    })
    .map((x) => x.r);
  const out: Placing[] = [];
  let place = 0;
  for (const r of sorted) {
    if (r.noTime) {
      out.push({ entryOrder: r.entryOrder, place: null, finalNs: 0, penaltyNs: penaltyNs(r) });
    } else {
      place += 1;
      out.push({ entryOrder: r.entryOrder, place, finalNs: finalNs(r), penaltyNs: penaltyNs(r) });
    }
  }
  return out;
}

/** Classifica dentro de cada categoria; a saída sai agrupada por categoria (ordem alfabética). */
export function rankByCategory(runs: ScoringRun[]): Placing[] {
  const cats: string[] = [];
  for (const r of runs) if (!cats.includes(r.category)) cats.push(r.category);
  cats.sort();
  const out: Placing[] = [];
  for (const cat of cats) out.push(...rank(runs.filter((r) => r.category === cat)));
  return out;
}
